### PR TITLE
feat: input table state

### LINKED ISSUE #29: Adaptivni input mapping
Nejdříve proběhne refactoring (#28), pak se to naimplementuje. Do input mappingu se bude předávat mapping state z docker runneru a input mapping si ještě k tomu načte table info a z toho vykoumá, co a jak má nastavit (v `getExportOptions`).

### PR DESCRIPTION
**Pull request by ondrejhlavacek**
FIXES #29 

- nové třídy `InputTablesState` a `InputTableState` (podobně jako `InputTablesOptions`)
- `InputTableState` drží v sobě název tabulky a jaký byl její `lastImportDate` při posledním importu
- `InputTableOptions::getStorageApiExportOptions` má jako argument `InputTableState` a pokud je v definici`"changed_since": "adaptive"`, tak `changed_since` přepíše na `InputTableState::getLastImportDate()`
- `Reader::downloadTables` vrací `InputTablesState`, která obsahuje aktuální hodnoty `lastImportDate` pro všechny zpracovávané tabulky. Hodnota `lastImportDate` se načte PŘED zafrontováním exportu. 

Implementace v Docker Runneru bude cca takhle

```php
$configuration = $this->client->getConfiguration('myComponent', 123456);
$tablesState = new InputTablesState($configuration['state']['storage']['input']['tables']);
$returnedTablesState = $reader->downloadTables($tablesDefinition, $tablesState, ...);
$this->client->updateConfiguration('myComponent', 123456, ['state' => $returnedTablesState->toArray());
```

1) načte se `InputTablesState` z konfigurace
2) předá se do `Reader::downloadTables`, která vrátí novou instanci `InputTablesState`
3) nový `InputTablesState` se uloží do state

### CODE REVIEW

**Comment by ondrejhlavacek** [comment]
Tak tohle je hodně weird

PHP 5.6

image


PHP 7.0

image

lokálně mi to samozřejmě jede

**Comment by odinuv** [comment]
to ma urcite nejaky uplne logicky vysvetleni

**Comment by ondrejhlavacek** [comment]
jj, zas nějakej sleep nebo něco takovýho

**Comment by ondrejhlavacek** [comment]
sleep to není. tohle bude oříšek, čuju nějakej problém v connection, blbý je, že lokálně to nejsem schopnej nasimulovat

**Comment by ondrejhlavacek** [comment]
@odinuv nepamatuješ si náhodou do jakýho projektu vede token na travisu?

**Comment by odinuv** [comment]
haha, zrovna jsem tenhle comment psal :D

**Comment by odinuv** [comment]
https://connection.keboola.com/admin/projects/4751/storage-explorer

**Comment by odinuv** [comment]
ale stejne to prochazi

**Comment by odinuv** [comment]
mozna neco specifickyho na php7.0 ?

**Comment by ondrejhlavacek** [comment]
dočasně jsem to na travisu omezil pouze na jeden test a nejede už ani jeden - https://travis-ci.org/keboola/input-mapping/builds/506177263

**Comment by odinuv** [comment]
no nechcu syckovat, ale vypada to, ze je v tom asi bug :) mam se do toho pustit nebo tomu jeste das?

**Comment by ondrejhlavacek** [comment]
jako jsem nápadama dost v koncích. zkusím přepnout default bucket projektu na snowflake, jestli tam nebude nějakej rozdíl (a tam se aspoň přes CLONE dostanu k `_timestamp`)

**Comment by odinuv** [comment]
si to udeleje explicitne v tom testu na swnoflakeu, kdyz prepnes default backend tak rozdrbes master testy

**Comment by ondrejhlavacek** [comment]
aha, jsem nevěděl, že to je takhle citlivý na backend, proč?

**Comment by odinuv** [comment]
no tak nemelo by byt - veci u kterejch je to citlivy se testuji explicitne na snowflake i redshiftu, ale stejne

**Comment by ondrejhlavacek** [comment]
lokálně mi to proti default snowflaku celý prošlo, zkouším ještě travis. a ta chyba je zdá se jen na redshiftu.

**Comment by ondrejhlavacek** [comment]
super, lokálně proti redshiftu taky chyba. ale teď teda nevím, co dál...

**Comment by ondrejhlavacek** [comment]
master na default snowflake prošel, nechám to tak

**Comment by ondrejhlavacek** [comment]
-> https://github.com/keboola/connection/issues/1881

Vypnu testování na Redshiftu a budu pokračovat.

File: Tests/ReaderTablesAdaptiveTest.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+        $tablesState = $reader->downloadTables($configuration, new InputTablesState([]), $this->temp->getTmpFolder() . DIRECTORY_SEPARATOR . "download");
+        $testTableInfo = $this->client->getTable("in.c-docker-test.test");
+        self::assertEquals(new \DateTime($testTableInfo['lastImportDate']), $tablesState->getTable("in.c-docker-test.test")->getLastImportDate());
+        self::assertCSVEquals(
+            "\"Id\",\"Name\",\"foo\",\"bar\"\n\"id1\",\"name1\",\"foo1\",\"bar1\"\n" .
+            "\"id2\",\"name2\",\"foo2\",\"bar2\"\n\"id3\",\"name3\",\"foo3\",\"bar3\"\n",
+            $this->temp->getTmpFolder() . DIRECTORY_SEPARATOR . "download/test.csv"
+        );
+        self::assertCount(1, $tablesState->toArray());
```
Comment (odinuv) [inline Tests/ReaderTablesAdaptiveTest.php]: je ten assert k necemu? ze je v tablessatate jedna tabulka se testuje uz tim prvnim assertem

File: Tests/ReaderTablesAdaptiveTest.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        ]);
+
+        // Create table
+        $csv = new CsvFile($this->temp->getTmpFolder() . DIRECTORY_SEPARATOR . "upload.csv");
+        $csv->writeRow(["Id", "Name", "foo", "bar"]);
+        $csv->writeRow(["id4", "name4", "foo4", "bar4"]);
+        $this->client->writeTableAsync("in.c-docker-test.test", $csv, ["incremental" => true]);
+
+        $this->client->exportTableAsync("in.c-docker-test.test");
+        $this->client->exportTableAsync("in.c-docker-test.test", ["changedSince" => $testTableInfo['lastImportDate']]);
```
Comment (odinuv) [inline Tests/ReaderTablesAdaptiveTest.php]: k cemu jsou ty ty exporty?

File: Tests/ReaderTablesAdaptiveTest.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+        self::assertEquals(new \DateTime($updatedTestTableInfo['lastImportDate']), $tablesState->getTable("in.c-docker-test.test")->getLastImportDate());
+        self::assertCSVEquals(
+            "\"Id\",\"Name\",\"foo\",\"bar\"\n\"id4\",\"name4\",\"foo4\",\"bar4\"\n",
+            $this->temp->getTmpFolder() . DIRECTORY_SEPARATOR . "download/test.csv"
+        );
+        self::assertCount(1, $tablesState->toArray());
+
+        $this->client->exportTableAsync("in.c-docker-test.test");
+        $this->client->exportTableAsync("in.c-docker-test.test", ["changedSince" => $testTableInfo['lastImportDate']]);
```
Comment (odinuv) [inline Tests/ReaderTablesAdaptiveTest.php]: dtto

File: Tests/ReaderTablesAdaptiveTest.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        // Create table
+        $csv = new CsvFile($this->temp->getTmpFolder() . DIRECTORY_SEPARATOR . "upload.csv");
+        $csv->writeRow(["Id", "Name", "foo", "bar"]);
+        $csv->writeRow(["id1", "name1", "foo1", "bar1"]);
+        $csv->writeRow(["id2", "name2", "foo2", "bar2"]);
+        $csv->writeRow(["id3", "name3", "foo3", "bar3"]);
+        $this->client->createTableAsync("in.c-docker-test", "test", $csv);
+    }
+
+    public function testDownloadTablesDownloadsTheWholeTable()
```
Comment (odinuv) [inline Tests/ReaderTablesAdaptiveTest.php]: tenhle test je podmnozina testu testDownloadTablesReturnsAllTablesTimestamps a je zbytecnej

File: Tests/ReaderTablesOutputStateTest.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    public function testDownloadTablesReturnsASingleTimestamps()
+    {
+        $reader = new Reader($this->client, new NullLogger());
+        $configuration = new InputTablesOptions([
+            [
+                "source" => "in.c-docker-test.test",
+                "destination" => "test.csv",
+            ]
+        ]);
+        $tablesState = $reader->downloadTables($configuration, new InputTablesState([]), $this->temp->getTmpFolder() . DIRECTORY_SEPARATOR . "download");
```
Comment (odinuv) [inline Tests/ReaderTablesOutputStateTest.php]: tohle je uplne zbytecnej test, testuje se to uz ve vsech ostatnich

File: Tests/ReaderTablesAdaptiveTest.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        $reader = new Reader($this->client, new NullLogger());
+        $configuration = new InputTablesOptions([
+            [
+                "source" => "in.c-docker-test.test",
+                "destination" => "test.csv",
+                "changed_since" => InputTableOptions::ADAPTIVE_INPUT_MAPPING_VALUE,
+            ]
+        ]);
+        $testTableInfo = $this->client->getTable("in.c-docker-test.test");
```
Comment (odinuv) [inline Tests/ReaderTablesAdaptiveTest.php]: imho by bylo lepsi 2x zavolat download tables, v druhym pouzit output z prvniho a mezi nima prihrat do tabulky data - bude to nejbliz realnymu pouziti - i kdyz to prakticky neznamena zmenu v testu, tak bude postavenej, tak jak to ma fungovat a bude to imho srozumitelnejsi

File: Reader/State/InputTablesState.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                return $table;
+            }
+        }
+        throw new TableNotFoundException('State for table "' . $tableName . '" not found.');
+    }
+
+    /**
+     * @return array
+     */
+    public function toArray()
```
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: použije to docker runner, aby to uložil do state

[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: k implementaci v docker runneru jsem se ještě nedostal, tam by to bylo vidět

File: Reader/State/InputTableState.php [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * @var \DateTime
+     */
+    private $lastImportDate;
+
+    public function __construct(array $configuration)
+    {
+        $this->source = $configuration['source'];
+        try {
+            $this->lastImportDate = new \DateTime($configuration['lastImportDate']);
```
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTableState.php]: máš pravdu, bejvávalo tam dřív nějaký porovnávání, ale to se mi podařilo eliminovat

File: Reader/State/InputTablesState.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Keboola\InputMapping\Reader\State;
+
+use Keboola\InputMapping\Reader\State\Exception\TableNotFoundException;
+
+class InputTablesState
```
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: super, dík!

File: Tests/ReaderTablesAdaptiveTest.php (hunk 1 above)
Comment (ondrejhlavacek) [reply Tests/ReaderTablesAdaptiveTest.php]: ten první assert to netestuje, jestli teda myslíš

```php
self::assertEquals(new \DateTime($testTableInfo['lastImportDate']), $tablesState->getTable("in.c-docker-test.test")->getLastImportDate());
```

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Reader/State/InputTablesState.php]: http://php.net/manual/en/class.jsonserializable.php ?

File: Tests/ReaderTablesAdaptiveTest.php (hunk 1 above)
Comment (odinuv) [reply Tests/ReaderTablesAdaptiveTest.php]: no kdyby tam ta tabulka nebyla, tak $tablesState->getTable("in.c-docker-test.test") nic nevrati

Comment (ondrejhlavacek) [reply Tests/ReaderTablesAdaptiveTest.php]: ano, ale kdyby tam byly třeba dvě, tak `$tablesState->getTable("in.c-docker-test.test")` něco vrátí a do state to uloží bordel.

Comment (odinuv) [reply Tests/ReaderTablesAdaptiveTest.php]: no budiz

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: aha, to je zajímavý, to jsem neznal! není tam nějakej protikus na deserializaci?

File: Reader/State/InputTableState.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTableState.php]: už možná vím - co když do toho někdo ručně zapíše nějakou píčovinu? takhle se aspoň zkontroluje, že tam je nějaký smysluplný datum. jinak to proboblá někam dál, ale to by asi hodilo zase error ze storage? vyzkouším!

[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTableState.php]:
```
Keboola\StorageApi\ClientException: Invalid date format: nonsense
```

to je ok, nechám to teda jako string

File: Reader/State/InputTablesState.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: mno, a mám rovnou přejmenovat `InputTablesOptions` na `InputTableOptionsList`?

[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: -> https://github.com/keboola/input-mapping/issues/31

File: Tests/InputTableStateTest.php [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        $state = new InputTableState(['source' => 'test', 'lastImportDate' => '2016-08-31T19:36:00+0200']);
+        self::assertEquals('test', $state->getSource());
+    }
+
+    public function testGetLastImportDate()
+    {
+        $state = new InputTableState(['source' => 'test', 'lastImportDate' => '2016-08-31T19:36:00+0200']);
+        self::assertEquals(new \DateTime('2016-08-31T19:36:00+0200'), $state->getLastImportDate());
+        self::assertEquals('2016', $state->getLastImportDate()->format('Y'));
```
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Tests/InputTableStateTest.php]: odstraněno

File: Tests/ReaderTablesAdaptiveTest.php (hunk 2 above)
Comment (ondrejhlavacek) [reply Tests/ReaderTablesAdaptiveTest.php]: tyjo, vůbec nevím. jdu se smazat, jestli se něco změní

Comment (ondrejhlavacek) [reply Tests/ReaderTablesAdaptiveTest.php]: test se asi o 50 vteřin zrychlil a nefailuje, takže dobrý :-)

File: Tests/ReaderTablesAdaptiveTest.php (hunk 6 above)
Comment (ondrejhlavacek) [reply Tests/ReaderTablesAdaptiveTest.php]: 👍

File: Reader/Reader.php [hunk 11]
```diff
@@ -272,9 +280,15 @@ public function downloadTables(InputTablesOptions $tablesDefinition, $destinatio
             } else {
                 throw new InvalidInputException("Parameter 'storage' must be either 'local' or 's3'.");
             }
+            $outputStateConfiguration[] = [
```
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/Reader.php]: jo, to je docela rozumný

File: Tests/ReaderTablesAdaptiveTest.php (hunk 4 above)
Comment (ondrejhlavacek) [reply Tests/ReaderTablesAdaptiveTest.php]: je to tak

File: Tests/ReaderTablesOutputStateTest.php (hunk 5 above)
Comment (ondrejhlavacek) [reply Tests/ReaderTablesOutputStateTest.php]: aha, a kdyby se tam třeba netestoval obsah CSV ale pouze ten state a naopek v `AdaptiveTest` se netestoval state a pouze obsah CSVček, bylo by to lepší? Radši bych udělal víc testů, který testujou různý části funkcionality, než jeden, kterej otestuje všechno naráz. Mohl by pak přijít bug/edge case, kterej to pak bude dost náročně větvit, když se to vyseparuje, bude to čistší.

File: Tests/InputTableStateTest.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Keboola\InputMapping\Tests;
+
+use Keboola\InputMapping\Reader\State\Exception\InvalidDateException;
+use Keboola\InputMapping\Reader\State\InputTableState;
+
+class InputTableStateTest extends \PHPUnit_Framework_TestCase
```
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Tests/InputTableStateTest.php]: jen nový věci, nebo mám rozházet namespace všeho? možná bych to teď nechal flat a dalším pullrequestem onamespacoval všechno? https://github.com/keboola/input-mapping/issues/31

**Comment by ondrejhlavacek** [comment]
resolvnul jsem všechny vyřešený věci a nechal tam pár nevyřešených na diskusi

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Reader/State/InputTablesState.php]: proc? json do php objektu jde prece serializovat vzdycky

File: Tests/InputTableStateTest.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Tests/InputTableStateTest.php]: no tak ty novy jsi tam mohl dat rovnou, ale je to asi fuk

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: jak?

File: Tests/ReaderTablesOutputStateTest.php (hunk 5 above)
Comment (odinuv) [reply Tests/ReaderTablesOutputStateTest.php]: stejne to bude duplikovany testem `testDownloadTablesReturnsAllTablesTimestamps`

> Mohl by pak přijít bug/edge case, kterej to pak bude dost náročně větvit

az prijde, tak se to vzdycky da predelat a cim min toho tam bude 2x, tim lip se to bude predelavat :)

File: Tests/InputTableStateTest.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Tests/InputTableStateTest.php]: nejsem si jistej, jestli by to pak při namespacování všeho dopadlo správně. ale udělám to jinak, namespacy a přejmenování na List udělám před tímhle pull requestem, ok?

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Reader/State/InputTablesState.php]: json_decode ?

[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: a jak to dostaneš do `InputTablesStateList`?

File: Tests/InputTableStateTest.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Tests/InputTableStateTest.php]: urcite, jestli ti to tak vyhovuje lip

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Reader/State/InputTablesState.php]: staticky ctor `fromData` kterej prijima vysledek json_decode

File: Tests/InputTableStateTest.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Tests/InputTableStateTest.php]: :+1:

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: `fromArray`? nebo to je něco jinýho?

File: Tests/ReaderTablesOutputStateTest.php (hunk 5 above)
Comment (ondrejhlavacek) [reply Tests/ReaderTablesOutputStateTest.php]: no jako přesně duplikovaný to není, chci otesetovat, že když jsou v bucketu 2 tabulky, ze který jednu naimportuješ, tak v state, který ti `Reader::downloadTables` vrátí, je pouze jedna tabulka (a to ta, kterou jsi stahoval, to tam ještě chybí). furt si myslíš, že to je duplikát?

File: Reader/State/InputTablesState.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (odinuv) [reply Reader/State/InputTablesState.php]: no nevim jestli to potrebujes na neco jinyho nez mas ted v konstructoru :) podle me ne, tak jsou to zbytecny uvahy. Jestli jedinej cil je mit moznost to serializovat do JSONu, tak bych tomu dal ten interface JsonSerializable, a konstruktor bude prijimat tu stejnou vec, kterou to vyprodukuje (patrne tj. bud stdclass nebo array, asi spis stdclass?)

takze dockerrunner zavola jsonserialize a o nic se nestara a pak posle zparsovanej json do ctoru a o nic se nestara.

[the remark this answers is not in the document]
Comment (ondrejhlavacek) [reply Reader/State/InputTablesState.php]: jo, teď jsem k tomu došel taky :-)

File: Tests/ReaderTablesOutputStateTest.php (hunk 5 above)
Comment (odinuv) [reply Tests/ReaderTablesOutputStateTest.php]: uch - jak s tim souvisi pocet tabulek v bucketu? ve state to ma vracet vzdycky tolik tabulek kolik jich bylo v konfiguraci, ne?

Comment (ondrejhlavacek) [reply Tests/ReaderTablesOutputStateTest.php]: přesně k tomu tam ten test je :-) když máš v bucketu 2 tabulky, v každý 2 řádky, v input mappingu 2 položky, každá s 2 parametrama (už přeháním), tak `2` na výstupu může bejt jakákoliv z těch hodnot. proto jsem chtěl udělat test, kde se něco takovýho stát nemůže a číslo ve výsledku je jednoznačný.

Comment (ondrejhlavacek) [reply Tests/ReaderTablesOutputStateTest.php]: matně si pamatuju, že mě jednou takováhle nejednoznačnost dost potrápila?

Comment (ondrejhlavacek) [reply Tests/ReaderTablesOutputStateTest.php]: NO DOBŘE!

**Comment by ondrejhlavacek** [comment]
rebasnul jsem po #32
